### ISSUE TITLE
Fails to preserve function __name__ (e.g. for decorators using @wraps)

### ISSUE THREAD

**Issue by jdanbrown** [cloudpipe/cloudpickle#177]
```py
>>> from functools import wraps
... import cloudpickle
...
... def f():
...     pass
...
... @wraps(f)
... def g():
...     f()

>>> f.__name__
'f'

>>> cloudpickle.loads(cloudpickle.dumps(f)).__name__
'f'

>>> g.__name__
'f'

>>> cloudpickle.loads(cloudpickle.dumps(g)).__name__
'g'
```

## Versions
```sh
$ python --version
Python 3.6.0
$ pip freeze | grep cloudpickle
cloudpickle==0.5.3
```

**Comment by Dveim** [comment]
Did you manage to find a workaround?

**Comment by jdanbrown** [comment]
@Dveim Not a general workaround, but here's what I've been using to hack around it for the purposes of using `joblib.Memory` within dask multiprocessing. IIRC the `func_name` vs. `__name__` hack only works because of the way `joblib.Memory.cache` looks for its cached function's name, but maybe this can spark some ideas for you.

```py
def wraps_workaround(f):
    """
    Workaround a cloudpickle bug that causes joblib.Memory to lose the cached func __name__ inside dask processes
    - https://github.com/cloudpipe/cloudpickle/issues/177

    This workaround can be deleted when these asserts pass:

        import dask
        from joblib.func_inspect import *

        load = Load()

        # Works regardless of bug, since threads don't invoke cloudpickle
        assert dask.delayed(format_signature)(load._metadata).compute(get=dask_get_for_scheduler_name('threads')) == \
            ('load._metadata', '_metadata()')

        # Requires bugfix to work, since processes marshal using cloudpickle
        assert dask.delayed(format_signature)(load._metadata).compute(get=dask_get_for_scheduler_name('processes')) == \
            ('load._metadata', '_metadata()')

    """
    def decorator(g):
        g = wraps(f)(g)
        g.func_name = f.__name__
        return g
    return decorator
```

**Comment by Dveim** [comment]
Thanks for response. Unfortunately, cannot imagine how to apply this to my case. Perhaps, will have to go with quick & dirty hacks in my code.

**Comment by llllllllll** [comment]
I believe that this is happening because we assume that the `co_name` and the function's `__name__` always match and the `__name__` doesn't appear in the vars of the function so the update doesn't get triggered.

**Comment by Dveim** [comment]
@llllllllll Wrote a hack in my code, then decided to check if I could fix this issue. Surprisingly, fix looks trivial, preparing PR.

**Comment by Dveim** [comment]
https://github.com/cloudpipe/cloudpickle/pull/183

If there are no issues with PR, could it be added to 0.5.4 release? (would like to have it as soon as possible)

**Comment by ogrisel** [comment]
I believe this was fixed in #183 but we forgot to close the issue.
